### PR TITLE
Support RESTful API module

### PR DESCRIPTION
**Pull request by yurabakhtin**
Issue: https://github.com/humhub/humhub-modules-rest/issues/33

### CODE REVIEW

File: config.php [hunk 1]
```diff
@@ -14,6 +15,7 @@
         ['class' => TopMenu::className(), 'event' => TopMenu::EVENT_INIT, 'callback' => ['humhub\modules\mail\Events', 'onTopMenuInit']],
         ['class' => NotificationArea::className(), 'event' => NotificationArea::EVENT_INIT, 'callback' => ['humhub\modules\mail\Events', 'onNotificationAddonInit']],
         ['class' => ProfileHeaderControls::className(), 'event' => ProfileHeaderControls::EVENT_INIT, 'callback' => ['humhub\modules\mail\Events', 'onProfileHeaderControlsInit']],
+        ['class' => Application::class, 'event' => Application::EVENT_BEFORE_REQUEST, 'callback' => ['humhub\modules\mail\Events', 'onBeforeRequest']],
```
Comment (luke-) [inline config.php:18]: Just as a thought: Could a separate new REST Module event be useful here?  I am not sure if we would have any advantages here. The event could e.g. collect the routes as Event attribute.

**Review by luke- — COMMENTED** [review]
Looks good for me. 
Do you plan to add the documentation and tests into this PR as well?

File: config.php (hunk 1 above)
Comment (yurabakhtin) [reply config.php:18]: Yes, my current solution doesn't require a new event trigger, and I find some modules already use the event `Application::EVENT_BEFORE_REQUEST` for other purposes. I need to investigate how to implement new separate event especially for the rest api url rules.

File: Events.php [hunk 2]
```diff
@@ -103,30 +103,30 @@ public static function onProfileHeaderControlsInit($event)
 
     public static function onRestApiAddRules()
     {
-        Yii::$app->urlManager->addRules([
+        Yii::$app->getModule('rest')->addRules([
```
Comment (luke-) [inline Events.php:106]: Can you please try to add something like: 
``` 
        /** @var \humhub\modules\rest\Module $module */
        $module = Yii::$app->getModule('rest');
```

Comment (yurabakhtin) [reply Events.php]: commit https://github.com/humhub/humhub-modules-mail/pull/217/commits/460e0915dad44a475603d904ffe6d12cf29ca669

File: config.php (hunk 1 above)
Comment (luke-) [reply config.php]: We don't necessarily have to create a new event class, it was just an idea. Currently we very rarely use custom event classes, but it's probably cleaner.

**Review by luke- — APPROVED** [review]

**Review by luke- — APPROVED** [review]
We should add `tests/codeception/_support/_generated` to ".gitignore"`?

**Comment by yurabakhtin** [comment]
@luke- The folder and its files were already ignored here https://github.com/humhub/humhub-modules-mail/blob/enh/33-rest-api-module/.gitignore#L9 and here https://github.com/humhub/humhub-modules-mail/blob/enh/33-rest-api-module/tests/codeception/_support/_generated/.gitignore#L2 but two files were added in repo by some unknown reason, so I just deleted them in commit https://github.com/humhub/humhub-modules-mail/pull/217/commits/b8ff3c78563c54379b1b7c4984c11a61dc16791f.

**Comment by luke-** [comment]
@yurabakhtin Ok great. Then we can finish the Mail module Rest API so far, or are there still points open?

**Pull request merged**